main.c

```diff
@@ -54,87 +104,173 @@ static struct bzimage_params initialize_from_fw_cfg()
+        linux_params->initrd_size = *(u32 *)boot_params.initrd_size_addr;
 
-        die_on(boot_params->initrd_start + boot_params->initrd_size
-                   > boot_params->initrd_addr_max,
+        die_on(linux_params->initrd_start + linux_params->initrd_size
+                   > linux_params->initrd_addr_max,
                "Initrd area exceeds maximum initrd address as required by the kernel\n");
     }
 
     const uint8_t custom_loader_type = 0xFF;
-    boot_params->loader_type = custom_loader_type;
+    linux_params->loader_type = custom_loader_type;
 
     // According to the Linux boot protocol, we should set up a simple GDT here. Both 32-bit
     // and 64-bit Linux set their own GDT right after the entry point, and simply jumping to
     // the entry address works without any issues so far, so we're skipping this.
 
-    uint8_t setup_sects = boot_params->setup_hdr;
+    uint8_t setup_sects = linux_params->setup_hdr;
 
     if (setup_sects == 0) {
         setup_sects = 4;
     }
 
-    size_t entry_ptr_32bit = bzimage_params.bzimage_addr + (setup_sects + 1) * 512;
+    size_t entry_ptr_32bit = boot_params.kernel_addr + (setup_sects + 1) * 512;
 
     // An overflowing unsigned integer addition will simply wrap. Such an overflow can be
     // detected by checking whether the result is smaller than one of the original values.
-    die_on(entry_ptr_32bit < bzimage_params.bzimage_addr,
-           "32bit entry pointer lies beyond 4G\n");
+    die_on(entry_ptr_32bit < boot_params.kernel_addr, "32bit entry pointer lies beyond 4G\n");
+
+    asm volatile("jmp *%[entry_ptr];" ::[entry_ptr] "r"(entry_ptr_32bit), "S"(linux_params),
+                 "m"(linux_params));
+
+    __builtin_unreachable();
+}
+
+// Boot an ELF binary according to the Multiboot specification.
+//
+// This boot method:
+//
+// 1. Extracts the ELF binary.
+// 2. Prepares Multiboot information (to pass the optional command line).
+// 3. Jumps to the extracted ELF's entry point.
+void elf_boot(struct boot_params params)
+{
+    die_on(
+        params.kernel_addr == 0,
+        "kernel start address not found!\n"
+        "The VMM does not offer an ELF via fw-cfg. Is the relevant config option missing?\n");
+
+    // Minimal (32bit) ELF loading.
+    struct elf_header *elf = (struct elf_header *)params.kernel_addr;
+    printf("Loading ELF from address: %p\n", elf);
+
+    bool is_elf_executable_or_unknown =
+        elf->type == ELF_TYPE_EXECUTABLE || elf->type == ELF_TYPE_UNKNOWN;
+    die_on(!is_elf_executable_or_unknown, "Unknown ELF type 0x%x!\n", elf->type);
+    die_on(elf->class == 2, "64bit ELF not yet supported!\n");
+    die_on(
+        elf->phentsize != 32,
+        "32bit ELF file should have program header entries with size 32, but found size %x!\n",
+        elf->phentsize);
+
+    // The ELF program header table describes how to prepare the memory layout.
+    printf("ELF program header table offset is 0x%x\n", elf->phoff);
```

Comment:
nit: I would limit the amout of printfs that the user sees by default. But YMMV.